microsoft/msix-packaging
# msixtest - use catch2 as our test infrastructure for all supported platforms

Before this change, we had three different implementations to perform our tests:
1. Powershell script, for windows
2. Shell script for Linux and macOS
3. Common shared library used for Android and iOS apps

This change removes the overhead of adding a test in three times by creating a single implementation by using Catch2.

msixtest is either an executable or a shared library, depending on the platform. It has a single entrypoint `msixtest_main` that takes argc and argv, as main, plus the path were the test packages are located. The shared library is used for our mobile test apps, while non-mobile just forwards the arguments to `msixtest_main`.

There is no custom command added, so msixtest has the same commands as Catch2. See [here](https://github.com/catchorg/Catch2/blob/master/docs/command-line.md)

Some usage examples:
msixtest : Runs all the tests. By default, catch2 doesn't show results of all success tests. It just shows failures and the number of test cases that passed.
msixtest -s : Runs all the tests and show results of all success tests
msixtest <test name> : Runs only the test specified
msixtest [tag] : Runs all tests in an specific tag, eg. [unpack] runs all the unpack tests.

Test:
- There are currently 66 tests.
- 1 test is specific for "no bundle".
- 2 tests are disabled, Api_AppxManifestReader_Capabilities and Unpack_SignedTamperedBlockMap-TRUST_E_BAD_DIGEST_sv. See comments on api_package.cpp and unpack.cpp for the reasons. 

Pipelines:
- Before this change, when the test task failed, we had to look into the logs and scroll until the tests are added. With these new tests we can use Catch2 reporters + Azure DevOps Test to provide a much better experience. All of our pipelines had been updated to pass `-r junit -o TEST-MsixSDK-$(_artifact).xml` to msixtest in the test task and a new publish test task to publish the result xml. I used junit because is a format that both Azure and Catch2 understand. 
- "No bundle" builds had been added to all platforms and are fully tested now. Before this change we didn't had an automated way to test "no bundle support"

Other related changes:
- Add Catch v2.7.2 license to the THIRD PARTY CODE NOTICE
- Move /test to /src/test to make it specific of the MSIX SDK codebase and to not be confused with MsixCore.
- Move samples binaries to their own output folder for clarity.
- Removes redundant BINARY_NAME CMake variable in CMakeLists.txt
- Make utf8_to_wstring and wstring_to_utf8 more strict for Windows.
- Fix bug were a file path with international characters couldn't be read on windows when CreateStreamOnFile is used
